android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/feature/BackStackFeature.kt

```diff
@@ -0,0 +1,193 @@
+                    op.configuration != state.current?.configuration -> {
+                        just(Effect.Replace(state, op.configuration))
+                    }
+                    else -> empty()
+                }
+
+                is Push -> when {
+                    op.configuration != state.current?.configuration ->
+                        just(Effect.Push(state, op.configuration))
+                    else -> empty()
+                }
+
+                is PushOverlay -> when {
+                    state.backStack.isNotEmpty() && op.configuration != state.currentOverlay ->
+                        just(Effect.PushOverlay(state, op.configuration))
+                    else -> empty()
+                }
+
+                is NewRoot -> when {
+                    state.backStack.size != 1 || state.backStack.first().configuration != op.configuration ->
+                        just(Effect.NewRoot(state, op.configuration))
+                    else -> empty()
+                }
+
+                is Pop -> when {
+                    state.canPopOverlay -> just(Effect.PopOverlay(state))
+                    state.canPopContent -> just(Effect.PopContent(state))
+                    else -> empty()
+                }
+            }
+    }
+
+    /**
+     * Creates a new [State] based on the old one + the applied [Effect]
+     */
+    @SuppressWarnings("LongMethod")
+    class ReducerImpl<C : Parcelable> : Reducer<BackStackFeatureState<C>, Effect<C>> {
+        override fun invoke(state: BackStackFeatureState<C>, effect: Effect<C>): BackStackFeatureState<C> =
+            state.apply(effect)
+
+        private fun BackStackFeatureState<C>.apply(effect: Effect<C>): BackStackFeatureState<C>  = when (effect) {
+            is Effect.Replace -> copy(
+                backStack = backStack.dropLast(1) + BackStackElement(effect.configuration)
+            )
+            is Effect.Push -> copy(
+                backStack = backStack + BackStackElement(effect.configuration)
+            )
+            is Effect.PushOverlay -> copy(
+                backStack = backStack.replaceLastWith {
+                    it.copy(
+                        overlays = it.overlays + effect.configuration
+                    )
+                }
+            )
+            is Effect.NewRoot -> copy(
+                backStack = listOf(BackStackElement(effect.configuration))
+            )
+            is Effect.PopOverlay -> copy(
+                backStack = backStack.dropLast(1) + backStack.last().copy(
+                    overlays = backStack.last().overlays.dropLast(1)
+                )
+            )
+            is Effect.PopContent -> copy(
+                backStack = backStack.dropLast(1)
+            )
+        }
+
+        private fun List<BackStackElement<C>>.replaceLastWith(replacement: (BackStackElement<C>) -> BackStackElement<C>): List<BackStackElement<C>> =
+            dropLast(1) + replacement.invoke(last())
```

Comment:
probably, `toMutableList().set(lastIndex, ...)` will be more efficient

Reply:
Done